### ISSUE TITLE
Capability providers segfault on initialization for Raspberry Pi

### ISSUE THREAD

**Issue by brooksmtownsend** [wasmCloud/wasmCloud#59]
Raspberry Pi 4 Model B Rev 1.2
```
OS: Raspbian GNU/Linux 10 (buster) armv7l
Arch: 32bit
Kernel: 5.4.72-v7l+
```

Currently experiencing Segmentation faults when scheduling capability providers on wasmCloud hosts running on raspberry pi's. This happens both on my RPI4 and my RPI3, they both have the same OS / Architecture pair. 

To reproduce:
1. Install `wasmcloud` via cargo, or by cloning the repository and building manually
2. Launch NATS on the pi
3. run wasmcloud with the arguments `RUST_LOG=info KVCACHE_NATS_URL=0.0.0.0:4222 wasmcloud`
4. Attempt to schedule a provider with `wash ctl start provider wasmcloud.azurecr.io/redis:0.9.2` (`redis`, `nats` and `httpserver` all fail)
5. segfault

These same parJEEzy's that are contained in the OCI registry work from my testing on the `x86_64-macos`, `x86-64-linux` and `aarch64-linux`, and each of the targets are built in the same fashion, using `cross` in our GH Actions CI/CD.

Additionally, I've attempted to build the `libwascc_redis.so` and form a parJEEzy on the PI to eliminate any architecture mismatches, and that also causes a segfault in the same initialization code.

This is the exact line of code that the program experiences a segfault: https://github.com/wasmCloud/wasmCloud/blob/main/crates/wasmcloud-host/src/capability/native_host.rs#L248. (just in case it changes later on, I'll paste the snippet here:
```
let plugin = unsafe {
    let constructor: Symbol<PluginCreate> = library.get(b"__capability_provider_create")?;
    let boxed_raw = constructor(); // RPI Segfault

    Box::from_raw(boxed_raw)
};
```

Looking for more information here, perhaps we need to do additional testing with the `constructor()` call for specific capability providers.

**Comment by autodidaddict** [comment]
Since every single provider fails during the call to `constructor()`, this feels like it might be build related or dispatch/invoke related, and not likely being caused by something inside the constructor itself. This is also supported by the fact that nearly all of the providers have virtually empty constructors (they do the bulk of their initialization _per actor_ during a `OP_BIND_ACTOR` handler).

**Comment by autodidaddict** [comment]
@billyoung Any thoughts on potential issues that might be occurring only with the 32-bit arm build of the shared objects that aren't happening with the other ones?

**Comment by brooksmtownsend** [comment]
@billyoung pointed out to me that it may be because I'm running the 32-bit version of raspbian on a pi 64-bit cpu. I will reflash one of my pi's to the 64bit version to see if this alleviates the issue, however it's still worth finding a native 32bit device to run the arm provider off of to make sure it's valid.

**Comment by billyoung** [comment]
I was unable to re-produce this issue :thinking:

System Info: ```$ uname -a
Linux rpi4 5.4.83-v7l+ #1379 SMP Mon Dec 14 13:11:54 GMT 2020 armv7l GNU/Linux```

Here are the steps I followed:

- Install system deps `sudo apt install libssl-dev clang`
- Install `sudo apt install docker docker-compose` for local dev
- Git clone `wasmCloud` and `wash`
- `cargo build` of `wasmCloud` and `wash`
- Start NATS `docker-compose -f ./hack/docker-compose.yml`

```
pi@rpi4:~/wasmCloud $ RUST_LOG=info KVCACHE_NATS_URL=0.0.0.0:4222 ./target/debug/wasmcloud
[2021-01-14T18:17:21Z INFO ] Message Bus started
[2021-01-14T18:17:21Z INFO ] Messagebus initialized
[2021-01-14T18:17:21Z INFO ] Messagebus initializing with lattice RPC support
[2021-01-14T18:17:21Z INFO ] Initializing lattice RPC client
[2021-01-14T18:17:21Z INFO ] Host Controller started
[2021-01-14T18:17:21Z INFO ] Native provider host started
[2021-01-14T18:17:21Z INFO ] Host controller initialized - NAFXDLVBNQYAZ322VXITKQAIYU2CS6OQOTZRYWJSEO76AUAJTCWAO4UK (Hot Updating - false)
[2021-01-14T18:17:21Z INFO ] Native provider host started
[2021-01-14T18:17:21Z INFO ] Creating lattice subscription for wasmbus://wasmcloud/keyvalue/__wasmcloud_lattice_cache/VAHNM37G4ARHZ3CYHB3L34M6TYQWQR6IZ4QVYC4NYZWTJCJ2LWP7S6Z2
[2021-01-14T18:17:21Z INFO ] Native Capability Provider 'wasmbus://wascc/extras/default/VDHPKGFKDI34Y4RN4PWWZHRYZ6373HYRSNNEM4UTDLLOGO5B37TSVREP' ready
[2021-01-14T18:17:21Z INFO ] Control Interface started
[2021-01-14T18:17:21Z INFO ] Native Capability Provider 'wasmbus://wasmcloud/keyvalue/__wasmcloud_lattice_cache/VAHNM37G4ARHZ3CYHB3L34M6TYQWQR6IZ4QVYC4NYZWTJCJ2LWP7S6Z2' ready
[2021-01-14T18:17:21Z INFO ] Initialized lattice cache provider

[2021-01-14T18:17:24Z INFO ] KV cache configured for system
[2021-01-14T18:17:24Z INFO ] Cache provider successfully configured
[2021-01-14T18:17:24Z INFO ] Host controller initialized - NAFXDLVBNQYAZ322VXITKQAIYU2CS6OQOTZRYWJSEO76AUAJTCWAO4UK (Hot Updating - false)
[2021-01-14T18:17:24Z INFO ] Initializing control interface - Active
[2021-01-14T18:20:02Z INFO ] Starting provider VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB
[2021-01-14T18:20:02Z INFO ] Native provider host started
[2021-01-14T18:20:02Z INFO ] Creating lattice subscription for wasmbus://wascc/keyvalue/default/VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB
[2021-01-14T18:20:02Z INFO ] Native Capability Provider 'wasmbus://wascc/keyvalue/default/VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB' ready
```
```
pi@rpi4:~/wash $ ./target/debug/wash ctl start provider wasmcloud.azurecr.io/redis:0.9.2
⡃⠀  Starting provider wasmcloud.azurecr.io/redis:0.9.2 ... 
Error starting provider: deadline has elapsed
```
```
pi@rpi4:~/wash $ ./target/debug/wash ctl get inventory NAFXDLVBNQYAZ322VXITKQAIYU2CS6OQOTZRYWJSEO76AUAJTCWAO4UK
⢋⠁  Retrieving inventory for host NAFXDLVBNQYAZ322VXITKQAIYU2CS6OQOTZRYWJSEO76AUAJTCWAO4UK ...

                        Host Inventory (NAFXDLVBNQYAZ322VXITKQAIYU2CS6OQOTZRYWJSEO76AUAJTCWAO4UK)                        
                                                                                                                         
 hostcore.osfamily                                           unix                                                        
 hostcore.arch                                               arm                                                         
 hostcore.os                                                 linux                                                       
                                                     No actors found                                                     
                                                                                                                         
 Provider ID                                                 Link Name                  Image Reference                  
 VDHPKGFKDI34Y4RN4PWWZHRYZ6373HYRSNNEM4UTDLLOGO5B37TSVREP    default                    N/A                              
 VAZVC4RX54J2NVCMCW7BPCAHGGG5XZXDBXFUMDUXGESTMQEJLC3YVZWB    default                    wasmcloud.azurecr.io/redis:0.9.2 
 VAHNM37G4ARHZ3CYHB3L34M6TYQWQR6IZ4QVYC4NYZWTJCJ2LWP7S6Z2    __wasmcloud_lattice_cache  N/A         
```

**Comment by autodidaddict** [comment]
So you're saying this is entirely Brooks's fault?

**Comment by billyoung** [comment]
I would never say such a thing @autodidaddict :rofl:

**Comment by brooksmtownsend** [comment]
Where's :party_try_not_to_cry: when you need it

**Comment by brooksmtownsend** [comment]
For me, this same segfault also occurs on Windows 10, using both the `x86_64-pc-windows-gnu` and `x86_64-pc-windows-msvc` targets at the exact same place in code. These two snippets are from the crash from the binary, and then the same crash in GDB with a bit more verbose location
```
[2021-01-15T17:34:13Z INFO ] constructor
error: process didn't exit successfully: `target\x86_64-pc-windows-gnu\release\wasmcloud.exe --rpc-host 127.0.0.1` (exit code: 0xc0000005, STATUS_ACCESS_VIOLATION)
```
```
[Switching to Thread 12028.0x467c]
0x00007ffbae2413ba in memmove () from C:\WINDOWS\SYSTEM32\vcruntime140.dll
```

In addition, I've also verified that the constructor() call should be successful within the (in this case) Redis capability provider. By putting `info!` calls around the provider initialization code, I was able to determine that it was not crashing while trying to call the `__capability_provider_create` function or when creating the `RedisKVProvider` struct. Narrowing the search down to:
```
let boxed_raw = constructor();
```
Something about this line, either assigning the pointer to the `boxed_raw` variable or the casting of `RedisKVProvider` to `dyn NativeCapabilityProvider` must be causing the segfault, sometimes, on arm-linux or x86_64-windows. Going to focus on the upcoming release for now, but we should revisit this when we can.

**Comment by brooksmtownsend** [comment]
Potentially fixed in #99

**Comment by brooksmtownsend** [comment]
This was fixed in #99! I'm no longer running into this issue on my rpi. cc @autodidaddict @billyoung, do we want to close this?

**Comment by autodidaddict** [comment]
:shipit:
